controllers/actions.summerwind.net/horizontal_runner_autoscaler_batch_scale.go

```diff
@@ -143,8 +143,51 @@ func (s *batchScaler) batchScale(ctx context.Context, batch batchScaleOperation)
 	}
 
 	copy := hra.DeepCopy()
+	now := time.Now()
+
+	if hra.Spec.MaxReplicas != nil && len(copy.Spec.CapacityReservations) > *hra.Spec.MaxReplicas {
+		// We have more reservations than MaxReplicas, meaning that we previously
+		// could not scale up to meet a capacity demand because we had hit MaxReplicas.
+		// Therefore, there are reservations that are starved for capacity. We extend the
+		// expiration time on these starved reservations because the "duration" is meant
+		// to apply to reservations that have launched replicas, not replicas in the backlog.
+		// Of course, if MaxReplicas is nil, then there is no max to hit, and we do not need this adjustment.
+		// See https://github.com/actions/actions-runner-controller/issues/2254 for more context.
+
+		// Extend the expiration time of all the reservations not yet assigned to replicas.
+		for i := *hra.Spec.MaxReplicas; i < len(copy.Spec.CapacityReservations); i++ {
+			// Let's say maxReplicas=3 and the workflow job of status=completed result in deleting the first capacity reservation
+			//   copy.Spec.CapacityReservations[i] where i=0.
+			// We are interested in at least four reservations and runners:
+			//   i=0   - already included in the current desired replicas, but may be about to be deleted
+			//   i=1-2 - already included in the current desired replicas
+			//   i=3   - not yet included in the current desired replicas, might have been expired while waiting in the queue
+			//
+			// i=3 is especially important here- If we didn't reset the expiration time of this reservation,
+			// it might expire before it is assigned to a runner, due to the delay between the time the
+			// expiration timer starts and the time a runner becomes available.
+			//
+			// Why is there such delay? Because ARC implements the scale duration and expiration as such...
+			// The expiration timer starts when the reservation is created, while the runner is created only after
+			// the corresponding reservation fits within maxReplicas.
+			//
+			// We address that, by resetting the expiration time for fourth(i=3 in the above example)
+			// and subsequent reservations whenever a batch is run (which is when expired reservations get deleted).
+
+			// There is no guarantee that all the reservations have the same duration, and even if there were,
+			// at this point we have lost the reference to the duration that was intended.
+			// However, we can compute the intended duration from the existing interval.
+			duration := copy.Spec.CapacityReservations[i].ExpirationTime.Time.Sub(copy.Spec.CapacityReservations[i].EffectiveTime.Time)
+			copy.Spec.CapacityReservations[i].EffectiveTime = metav1.Time{Time: now}
+			copy.Spec.CapacityReservations[i].ExpirationTime = metav1.Time{Time: now.Add(duration)}
```

Comment:
@Nuru Can I take it as this line corresponds to the following description?

> Because reservations could have been added by multiple triggers with different durations, duration is calculated from the existing expiration and effective times rather than by the duration of whatever scale-down trigger has fired.

We've previously used the "latest capacity reservation duration" for the HRA contained in the scaleTrigger object. You made it keep using the capacity reservation duration at the time of the capacity reservation was initially added, thinking this is more correct than the previous logic. Am I getting it correctly?

Reply:
Yes. The expiration duration is set when the scale-up event occurs and should not change. Plus, you could (in theory, anyway) have multiple triggers with different duration configurations acting at the same time and you would not want them to interfere with each other.

Reply:
> multiple triggers with different duration configurations acting at the same time

Acting at the same time against the same HorizontalRunnerAutoscaler, right? 🤔
I don't think that is supposed to work anyway, because a workflow job is routed to only one eligible HRA.
Anyway, making it to reuse the duration sounds good so I'm definitely going to keep this one!
Thanks for your confirmation.